nginx-odoo.py

```diff
@@ -8,326 +8,169 @@
+            message = 'Mail with security code not sent.'
+            logging.error(message)
+            return self.render(r'./templates/login.html',**config.theme_params,error=message)
+        return self.render(r'./templates/hotp.html',**config.theme_params,counter=counter,code=code)
+      else:
+        # TODO: brute force protection
+        #       (block for X minutes after X attempts)
+        message = 'Invalid username or password.'
+        logging.info(message)
+        return self.render(r'./templates/login.html',**config.theme_params,error=message)
 
     # check HOTP
-    counter = request.forms.get('counter')
-    code = request.forms.get('code')
-    hotp_code = request.forms.get('hotp_code')
+    counter=self.get_body_argument('counter',default=None)
+    code=self.get_body_argument('code',default=None)
+    hotp_code=self.get_body_argument('hotp_code',default=None)
     if code and counter and hotp_code:
-        hotp = pyotp.HOTP(HOTP_SECRET)
-        if not hotp.verify(hotp_code, int(counter)):
-            message = 'Invalid security code.'
-            return template('login', dict(theme_params, error=message))
-        session_id = db.verify_code_and_expiry(counter, code)
-        if not session_id:
-            message = 'Invalid security code (2).'
-            return template('login', dict(theme_params, error=message))
-        db.save_session(session_id, EXPIRY_INTERVAL)
-        logging.info('Setting session cookie: %s', session_id)
-        response.set_cookie("session_id", session_id, path='/')
-        return redirect('/')
-
-    return redirect('/')
-
-class StripPathMiddleware(object):
-    '''
-    Get that slash out of the request
-    '''
-    def __init__(self, a):
-        self.a = a
-    def __call__(self, e, h):
-        e['PATH_INFO'] = e['PATH_INFO'].rstrip('/')
-        return self.a(e, h)
-
+      hotp = pyotp.HOTP(config.HOTP_SECRET)
+      if not hotp.verify(hotp_code, int(counter)):
+        message = 'Invalid security code.'
+        return self.render(r'./templates/login.html',**config.theme_params,error=message)
+      session_id = db.verify_code_and_expiry(counter, code)
+      if not session_id:
+        message = 'Invalid security code (2).'
+        return self.render(r'./templates/login.html',**config.theme_params,error=message)
+      db.save_session(session_id, config.EXPIRY_INTERVAL)
+      logging.info('Setting session cookie: %s', session_id)
+      self.set_cookie('session_id',session_id,path='/')
+      # Redirect to other service
+      # needed when using the bouncer to authenticate a user with the bouncer
+      redirect_url=self.get_argument('redirect','/')
+      if redirect_url!='/':
+        if not redirect_url.endswith('/'):
+          redirect_url+='/'
+        redirect_url=f'{redirect_url}auth/{session_id}'
+      return self.redirect(redirect_url)
+    return self.redirect('/')
+
+# Session verificaiton
+class VerifySessionHandler(RequestHandler):
+  def get(self):
```

Comment:
It would be good if this could also become async, because this function is called very often during normal operation, basically for every HTTP request that comes into Odoo